<a href="https://colab.research.google.com/github/harit77/deep-learn/blob/main/chapter02_mathematical-building-blocks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This is a companion notebook for the book [Deep Learning with Python, Third Edition](https://www.manning.com/books/deep-learning-with-python-third-edition). For readability, it only contains runnable code blocks and section titles, and omits everything else in the book: text paragraphs, figures, and pseudocode.

**If you want to be able to follow what's going on, I recommend reading the notebook side by side with your copy of the book.**

The book's contents are available online at [deeplearningwithpython.io](https://deeplearningwithpython.io).

In [ ]:
!pip install keras keras-hub --upgrade -q

In [ ]:
import os
os.environ["KERAS_BACKEND"] = "tensorflow"

In [ ]:
# @title
import os
from IPython.core.magic import register_cell_magic

@register_cell_magic
def backend(line, cell):
    current, required = os.environ.get("KERAS_BACKEND", ""), line.split()[-1]
    if current == required:
        get_ipython().run_cell(cell)
    else:
        print(
            f"This cell requires the {required} backend. To run it, change KERAS_BACKEND to "
            f"\"{required}\" at the top of the notebook, restart the runtime, and rerun the notebook."
        )

## The mathematical building blocks of neural networks

### A first look at a neural network

In [1]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


In [2]:
train_images.shape

(60000, 28, 28)

In [3]:
len(train_labels)

60000

In [17]:
train_labels.ndim

1

In [18]:
test_images.shape

(10000, 28, 28)

In [19]:
len(test_labels)

10000

In [20]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

In [21]:
import keras
from keras import layers

model = keras.Sequential(
    [
        layers.Dense(512, activation="relu"),
        layers.Dense(10, activation="softmax"),
    ]
)

In [22]:
model.compile(
    optimizer="adam",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"],
)

In [23]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

In [24]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.9252 - loss: 0.2637
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9683 - loss: 0.1082
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9800 - loss: 0.0681
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9853 - loss: 0.0499
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9890 - loss: 0.0374


In [25]:
test_digits = test_images[0:10]
predictions = model.predict(test_digits)
predictions[0]

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 386ms/step


array([4.1426344e-07, 1.8861519e-08, 5.1915013e-05, 2.8122056e-04,
       7.2917311e-10, 4.9362450e-07, 3.1528080e-10, 9.9962842e-01,
       2.3272452e-05, 1.4334939e-05], dtype=float32)

In [26]:
predictions[0].argmax()

np.int64(7)

In [27]:
predictions[0][7]

np.float32(0.9996284)

In [32]:
test_labels[0]

np.uint8(7)

In [33]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"test_acc: {test_acc}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9800 - loss: 0.0623
test_acc: 0.9800000190734863


### Comparing Training and Test Accuracy

In [34]:
print(f"Training accuracy (from last epoch): {0.9890:.4f}")
print(f"Test accuracy: {test_acc:.4f}")

Training accuracy (from last epoch): 0.9890
Test accuracy: 0.9800


### Data representations for neural networks

#### Scalars (rank-0 tensors)

In [ ]:
import numpy as np
x = np.array(12)
x

In [ ]:
x.ndim

#### Vectors (rank-1 tensors)

In [ ]:
x = np.array([12, 3, 6, 14, 7])
x

In [ ]:
x.ndim

#### Matrices (rank-2 tensors)

In [ ]:
x = np.array([[5, 78, 2, 34, 0],
              [6, 79, 3, 35, 1],
              [7, 80, 4, 36, 2]])
x.ndim

#### Rank-3 tensors and higher-rank tensors

In [ ]:
x = np.array([[[5, 78, 2, 34, 0],
               [6, 79, 3, 35, 1],
               [7, 80, 4, 36, 2]],
              [[5, 78, 2, 34, 0],
               [6, 79, 3, 35, 1],
               [7, 80, 4, 36, 2]],
              [[5, 78, 2, 34, 0],
               [6, 79, 3, 35, 1],
               [7, 80, 4, 36, 2]]])
x.ndim

#### Key attributes

In [ ]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [ ]:
train_images.ndim

In [ ]:
train_images.shape

In [ ]:
train_images.dtype

In [ ]:
import matplotlib.pyplot as plt

digit = train_images[4]
plt.imshow(digit, cmap=plt.cm.binary)
plt.show()

In [ ]:
train_labels[4]

#### Manipulating tensors in NumPy

In [37]:
my_slice_both_dimensions = train_images[0:100, 0:28]
my_slice_both_dimensions.shape

(100, 28)

In [39]:
my_slice = train_images[10:100]
my_slice.shape

(90, 784)

In [40]:
my_slice = train_images[10:100, :, :]
my_slice.shape

IndexError: too many indices for array: array is 2-dimensional, but 3 were indexed

In [ ]:
my_slice = train_images[10:100, 0:28, 0:28]
my_slice.shape

In [ ]:
my_slice = train_images[:, 14:, 14:]

In [ ]:
my_slice = train_images[:, 7:-7, 7:-7]

#### The notion of data batches

In [ ]:
batch = train_images[:128]

In [ ]:
batch = train_images[128:256]

In [ ]:
n = 3
batch = train_images[128 * n : 128 * (n + 1)]

#### Real-world examples of data tensors

##### Vector data

##### Timeseries data or sequence data

##### Image data

##### Video data

### The gears of neural networks: Tensor operations

#### Element-wise operations

In [ ]:
def naive_relu(x):
    assert len(x.shape) == 2
    x = x.copy()
    for i in range(x.shape[0]):
        for j in range(x.shape[1]):
            x[i, j] = max(x[i, j], 0)
    return x

In [ ]:
def naive_add(x, y):
    assert len(x.shape) == 2
    assert x.shape == y.shape
    x = x.copy()
    for i in range(x.shape[0]):
        for j in range(x.shape[1]):
            x[i, j] += y[i, j]
    return x

In [ ]:
import time

x = np.random.random((20, 100))
y = np.random.random((20, 100))

t0 = time.time()
for _ in range(1000):
    z = x + y
    z = np.maximum(z, 0.0)
print("Took: {0:.2f} s".format(time.time() - t0))

In [ ]:
t0 = time.time()
for _ in range(1000):
    z = naive_add(x, y)
    z = naive_relu(z)
print("Took: {0:.2f} s".format(time.time() - t0))

#### Broadcasting

In [ ]:
import numpy as np

X = np.random.random((32, 10))
y = np.random.random((10,))

In [ ]:
y = np.expand_dims(y, axis=0)

In [ ]:
Y = np.tile(y, (32, 1))

In [ ]:
def naive_add_matrix_and_vector(x, y):
    assert len(x.shape) == 2
    assert len(y.shape) == 1
    assert x.shape[1] == y.shape[0]
    x = x.copy()
    for i in range(x.shape[0]):
        for j in range(x.shape[1]):
            x[i, j] += y[j]
    return x

In [ ]:
import numpy as np

x = np.random.random((64, 3, 32, 10))
y = np.random.random((32, 10))
z = np.maximum(x, y)

#### Tensor product

In [ ]:
x = np.random.random((32,))
y = np.random.random((32,))

z = np.matmul(x, y)
z = x @ y

In [ ]:
def naive_vector_product(x, y):
    assert len(x.shape) == 1
    assert len(y.shape) == 1
    assert x.shape[0] == y.shape[0]
    z = 0.0
    for i in range(x.shape[0]):
        z += x[i] * y[i]
    return z

In [ ]:
def naive_matrix_vector_product(x, y):
    assert len(x.shape) == 2
    assert len(y.shape) == 1
    assert x.shape[1] == y.shape[0]
    z = np.zeros(x.shape[0])
    for i in range(x.shape[0]):
        for j in range(x.shape[1]):
            z[i] += x[i, j] * y[j]
    return z

In [ ]:
def naive_matrix_vector_product(x, y):
    z = np.zeros(x.shape[0])
    for i in range(x.shape[0]):
        z[i] = naive_vector_product(x[i, :], y)
    return z

In [ ]:
def naive_matrix_product(x, y):
    assert len(x.shape) == 2
    assert len(y.shape) == 2
    assert x.shape[1] == y.shape[0]
    z = np.zeros((x.shape[0], y.shape[1]))
    for i in range(x.shape[0]):
        for j in range(y.shape[1]):
            row_x = x[i, :]
            column_y = y[:, j]
            z[i, j] = naive_vector_product(row_x, column_y)
    return z

#### Tensor reshaping

In [ ]:
train_images = train_images.reshape((60000, 28 * 28))

In [ ]:
x = np.array([[0., 1.],
              [2., 3.],
              [4., 5.]])
x.shape

In [ ]:
x = x.reshape((6, 1))
x

In [ ]:
x = x.reshape((2, 3))
x

In [ ]:
x = np.zeros((300, 20))
x = np.transpose(x)
x.shape

#### Geometric interpretation of tensor operations

#### A geometric interpretation of deep learning

### The engine of neural networks: Gradient-based optimization

#### What's a derivative?

#### Derivative of a tensor operation: The gradient

#### Stochastic gradient descent

#### Chaining derivatives: The Backpropagation algorithm

##### The chain rule

##### Automatic differentiation with computation graphs

### Looking back at our first example

In [ ]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

In [ ]:
model = keras.Sequential(
    [
        layers.Dense(512, activation="relu"),
        layers.Dense(10, activation="softmax"),
    ]
)

In [ ]:
model.compile(
    optimizer="adam",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"],
)

In [ ]:
model.fit(
    train_images,
    train_labels,
    epochs=5,
    batch_size=128,
)

After these 5 epochs, the model will have performed 2,345 gradient updates (469 per epoch), and the loss of the model will be sufficiently low that the model will be capable of classifying handwritten digits with high accuracy." Explain this it from the book. I din't understand how did we reach to 469
That's a great question, and it's all about how the model processes the training data in 'batches'!

Here's how we get to 469:

Total Training Samples: From the mnist.load_data() call, we know train_images contains 60,000 training examples (digits).
Batch Size: When the model is trained with model.fit(..., batch_size=128), it processes the data in chunks of 128 images at a time. Each chunk is called a 'batch'.
Gradient Update per Batch: For each batch, the model calculates the loss, then computes the gradients (how much each weight should change) and updates its weights. This is one 'gradient update'.
To figure out how many batches (and thus gradient updates) happen in one full pass over the entire training set (one epoch), you divide the total number of training samples by the batch size:

60,000 (total training images) / 128 (batch size) = 468.75

Since you can't have a fraction of a batch, the last batch will simply be smaller than 128. The model still performs a gradient update for this partial batch. Therefore, the number of gradient updates per epoch is rounded up to 469

#### Reimplementing our first example from scratch

##### A simple Dense class

In [42]:
import keras
from keras import ops

class NaiveDense:
    def __init__(self, input_size, output_size, activation=None):
        self.activation = activation
        self.W = keras.Variable(
            shape=(input_size, output_size), initializer="uniform"
        )
        self.b = keras.Variable(shape=(output_size,), initializer="zeros")

    def __call__(self, inputs):
        x = ops.matmul(inputs, self.W)
        x = x + self.b
        if self.activation is not None:
            x = self.activation(x)
        return x

    @property
    def weights(self):
        return [self.W, self.b]

This code defines a NaiveDense class, which is a custom implementation of a fully connected (dense) layer, similar to what you'd find in a deep learning library like Keras. It's designed to illustrate the fundamental operations within such a layer without relying on Keras's built-in Dense layer.

Here's a breakdown:

import keras and from keras import ops: These lines import the Keras library and its ops module, which provides backend-agnostic tensor operations.

class NaiveDense:: This defines the custom dense layer class.

__init__(self, input_size, output_size, activation=None): This is the constructor for the NaiveDense layer.

input_size: Specifies the number of input features for this layer.
output_size: Specifies the number of output units (neurons) in this layer.
activation: An optional activation function (e.g., relu, softmax) to apply to the output. If None, no activation is applied.
self.W: This creates the weight matrix using keras.Variable. It has a shape of (input_size, output_size) and is initialized with random uniform values. These are the trainable parameters that transform the input.
self.b: This creates the bias vector, also a keras.Variable. It has a shape of (output_size,) and is initialized with zeros. The bias is added to the output of the matrix multiplication.
__call__(self, inputs): This method defines the forward pass of the layer, i.e., how it processes input data.

x = ops.matmul(inputs, self.W): This performs a matrix multiplication of the inputs with the weight matrix self.W. This is the core linear transformation.
x = x + self.b: The bias vector self.b is added to the result of the matrix multiplication. This is an element-wise addition, and broadcasting rules apply if the shapes don't perfectly match.
if self.activation is not None: x = self.activation(x): If an activation function was provided during initialization, it's applied to the result. Activation functions introduce non-linearity into the model.
return x: The processed output of the layer is returned.
@property def weights(self): return [self.W, self.b]: This defines a property named weights that allows you to easily access the trainable parameters of the layer (the weight matrix W and the bias vector b). This is useful when you need to collect all trainable variables from a model, for example, during optimization.

##### A simple Sequential class

In [43]:
class NaiveSequential:
    def __init__(self, layers):
        self.layers = layers

    def __call__(self, inputs):
        x = inputs
        for layer in self.layers:
            x = layer(x)
        return x

    @property
    def weights(self):
        weights = []
        for layer in self.layers:
            weights += layer.weights
        return weights

Certainly! This code defines a NaiveSequential class, which is a custom, simplified implementation of a sequential model, much like keras.Sequential. It allows you to stack layers together to create a neural network model.

Here's a breakdown:

class NaiveSequential:: This defines the custom sequential model class.

def __init__(self, layers):: This is the constructor for the NaiveSequential model.

It takes a list of layers as an argument. These layers are expected to be callable objects (like the NaiveDense layers you might define or Keras layers) that can process inputs.
self.layers = layers: It simply stores this list of layers as an attribute of the NaiveSequential instance.
def __call__(self, inputs):: This method defines the forward pass of the entire sequential model.

x = inputs: It initializes x with the initial inputs to the model.
for layer in self.layers:: It then iterates through each layer in the list of layers.
x = layer(x): For each layer, it calls the layer with the current x (which is the output of the previous layer or the initial input). The output of the current layer becomes the input for the next layer.
return x: Finally, it returns the output of the last layer, which is the model's prediction.
@property def weights(self):: This defines a property called weights that allows you to easily collect all trainable weights from all layers within the sequential model.

weights = []: It initializes an empty list to store the weights.
for layer in self.layers:: It iterates through each layer in the model.
weights += layer.weights: It assumes each layer itself has a weights property (or method) that returns its own list of weights, and it extends the weights list with these.
return weights: It returns the comprehensive list of all trainable weights from all layers in the model. This is crucial for optimization algorithms that need to update all model parameters.

In [44]:
model = NaiveSequential(
    [
        NaiveDense(input_size=28 * 28, output_size=512, activation=ops.relu),
        NaiveDense(input_size=512, output_size=10, activation=ops.softmax),
    ]
)
assert len(model.weights) == 4

This code block initializes a neural network model using the custom NaiveSequential class, which acts like a container for stacking layers sequentially. Here's a breakdown:

model = NaiveSequential(...): This line creates an instance of your NaiveSequential model.

First NaiveDense layer: NaiveDense(input_size=28 * 28, output_size=512, activation=ops.relu)

This is the first layer of the network. It's a fully connected (dense) layer.
input_size=28 * 28: This specifies that the input to this layer will be flattened images, each with 28 * 28 = 784 pixels. This is typical for processing MNIST handwritten digit images which are 28x28 pixels.
output_size=512: This means this layer will produce 512 output features.
activation=ops.relu: The Rectified Linear Unit (ReLU) activation function is applied to the output of this layer, introducing non-linearity into the model.
Second NaiveDense layer: NaiveDense(input_size=512, output_size=10, activation=ops.softmax)

This is the second and final layer of the network.
input_size=512: The input to this layer comes from the 512 output features of the previous layer.
output_size=10: This layer outputs 10 values, corresponding to the 10 possible digit classes (0 through 9) in the MNIST dataset.
activation=ops.softmax: The softmax activation function is used here. It converts the raw outputs into a probability distribution over the 10 classes, where each output is a probability that the input image belongs to a specific digit class, and all probabilities sum to 1.
assert len(model.weights) == 4:

This line is an assertion to check if the model has the expected number of trainable weights.
Each NaiveDense layer has two sets of weights: a weight matrix (W) and a bias vector (b).
Since there are two NaiveDense layers in this NaiveSequential model, the total number of weight tensors (W1, b1, W2, b2) is 4. This assertion confirms that the custom implementation correctly tracks these parameters.

 The Rectified Linear Unit (ReLU) activation function is applied to the output of this layer, introducing non-linearity into the model." what do mean by non-linearity and why its important
That's a great question! Non-linearity is a fundamental concept in neural networks, and it's what allows them to learn complex patterns.

What is Non-linearity?
In the context of activation functions like ReLU, "non-linearity" means that the function's output is not directly proportional to its input. If you were to plot a linear function (like y = mx + b), it would be a straight line. A non-linear function, however, can produce curves, bends, and other more intricate shapes.

For example, the ReLU function f(x) = max(0, x) is non-linear because for x < 0, the output is 0, and for x >= 0, the output is x. This creates a 'bend' or 'kink' at x=0, making it not a straight line across its entire domain.

Why is Non-linearity Important?
If all activation functions in a neural network were linear, then regardless of how many layers you stacked, the entire network would behave just like a single linear layer. Let's say you have two linear layers: y = W1 * x + b1 and z = W2 * y + b2. If you substitute y into the second equation, you get z = W2 * (W1 * x + b1) + b2, which simplifies to z = (W2 * W1) * x + (W2 * b1 + b2). This is still just a linear transformation of x (let W_new = W2 * W1 and b_new = W2 * b1 + b2, then z = W_new * x + b_new).

This means a network with only linear layers could only learn linear relationships between input and output. It would be incapable of modeling anything more complex than a straight line or a simple plane. Most real-world data and tasks, however, involve highly complex, non-linear relationships (e.g., recognizing images, understanding speech, predicting stock prices).

Non-linear activation functions are crucial because they:

Enable Learning Complex Patterns: They introduce the ability for the network to approximate any arbitrary function, allowing it to learn intricate and non-linear mappings from inputs to outputs. Without non-linearity, deep neural networks would be no more powerful than simple linear regression models.
Allow for Deep Architectures: They are what makes "deep" learning possible. By stacking multiple non-linear layers, the network can learn hierarchical representations of the data, extracting progressively more abstract and useful features at each layer.
In essence, non-linearity gives neural networks the expressive power to model the complexities of the real world.

##### A batch generator

In [ ]:
import math

class BatchGenerator:
    def __init__(self, images, labels, batch_size=128):
        assert len(images) == len(labels)
        self.index = 0
        self.images = images
        self.labels = labels
        self.batch_size = batch_size
        self.num_batches = math.ceil(len(images) / batch_size)

    def next(self):
        images = self.images[self.index : self.index + self.batch_size]
        labels = self.labels[self.index : self.index + self.batch_size]
        self.index += self.batch_size
        return images, labels

#### Running one training step

##### The weight update step

In [ ]:
learning_rate = 1e-3

def update_weights(gradients, weights):
    for g, w in zip(gradients, weights):
        w.assign(w - g * learning_rate)

In [ ]:
from keras import optimizers

optimizer = optimizers.SGD(learning_rate=1e-3)

def update_weights(gradients, weights):
    optimizer.apply_gradients(zip(gradients, weights))

##### Gradient computation

In [ ]:
%%backend tensorflow
import tensorflow as tf

x = tf.zeros(shape=())
with tf.GradientTape() as tape:
    y = 2 * x + 3
grad_of_y_wrt_x = tape.gradient(y, x)

In [ ]:
%%backend tensorflow
def one_training_step(model, images_batch, labels_batch):
    with tf.GradientTape() as tape:
        predictions = model(images_batch)
        loss = ops.sparse_categorical_crossentropy(labels_batch, predictions)
        average_loss = ops.mean(loss)
    gradients = tape.gradient(average_loss, model.weights)
    update_weights(gradients, model.weights)
    return average_loss

#### The full training loop

In [ ]:
%%backend tensorflow
def fit(model, images, labels, epochs, batch_size=128):
    for epoch_counter in range(epochs):
        print(f"Epoch {epoch_counter}")
        batch_generator = BatchGenerator(images, labels)
        for batch_counter in range(batch_generator.num_batches):
            images_batch, labels_batch = batch_generator.next()
            loss = one_training_step(model, images_batch, labels_batch)
            if batch_counter % 100 == 0:
                print(f"loss at batch {batch_counter}: {loss:.2f}")

In [ ]:
%%backend tensorflow
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

fit(model, train_images, train_labels, epochs=10, batch_size=128)

#### Evaluating the model

In [ ]:
%%backend tensorflow
predictions = model(test_images)
predicted_labels = ops.argmax(predictions, axis=1)
matches = predicted_labels == test_labels
f"accuracy: {ops.mean(matches):.2f}"